src/org/sosy_lab/java_smt/basicimpl/FallbackOptimizationProver.java

```diff
@@ -0,0 +1,203 @@
+        this.baseProver = pContext.newProverEnvironment(pOptions);
+        this.objectives = new HashMap<>();
+        this.nextHandle = 0;
+    }
+    
+    @Override
+    public boolean isOptimizationSupported() {
+        return true; // We always support optimization through fallback
+    }
+    
+    @Override
+    protected String getSolverName() {
+        return context.getSolverName();
+    }
+    
+    @Override
+    protected int maximizeInternal(Formula objective) {
+        int handle = nextHandle++;
+        objectives.put(handle, objective);
+        return handle;
+    }
+    
+    @Override
+    protected int minimizeInternal(Formula objective) {
+        // For minimization, we negate the objective and maximize
+        FormulaManager fmgr = context.getFormulaManager();
+        if (objective instanceof IntegerFormula) {
+            IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();
+            return maximizeInternal(imgr.negate((IntegerFormula) objective));
+        } else if (objective instanceof RationalFormula) {
+            // Handle rational formulas similarly
+            // Implementation depends on the specific formula manager
+            throw new UnsupportedOperationException(
+                "Rational optimization not yet implemented in fallback");
+        } else {
+            throw new UnsupportedOperationException(
+                "Unsupported formula type for optimization: " + objective.getClass());
+        }
+    }
+    
+    @Override
+    protected Optional<Rational> upperInternal(int handle, Rational epsilon) {
+        Formula objective = objectives.get(handle);
+        if (objective == null) {
+            throw new IllegalArgumentException("Invalid objective handle: " + handle);
+        }
+        
+        try {
+            // Start with a large upper bound
+            Rational upperBound = Rational.ofLong(1000000);
+            Rational lowerBound = Rational.ofLong(-1000000);
+            
+            // Binary search to find the maximum value
+            while (upperBound.subtract(lowerBound).compareTo(epsilon) > 0) {
+                Rational mid = upperBound.add(lowerBound).divide(Rational.ofLong(2));
+                
+                // Create constraint: objective <= mid
+                BooleanFormula constraint = createUpperBoundConstraint(objective, mid);
+                baseProver.push();
+                baseProver.addConstraint(constraint);
+                
+                try {
+                    if (baseProver.isUnsat()) {
+                        // No solution exists with this upper bound
+                        upperBound = mid;
+                    } else {
+                        // Solution exists, try a higher value
+                        lowerBound = mid;
+                    }
+                } finally {
+                    baseProver.pop();
+                }
+            }
+            
+            return Optional.of(upperBound);
```

Comment:
Binary search is a valid idea and it should work in most applications. As a first prototype, this might work for several use cases. The performance overhead is not that large, e.g., several SAT checks are required (which is ok) and the number of them is logarithmic in the bounding window size (which is good).

There are some open points:
- We need to think about upper/lower bounds on unbounded theories like Integer or Rational.
- Maybe keep the prover content if the next contraint it stricter than the previous. This might improve reuse and perfomance on larger queries.